core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -430,7 +430,7 @@ sealed trait ZIO[-R, +E, +A] extends Serializable with ZIOPlatformSpecific[R, E,
       for {
         id    <- ZIO.fiberId
         fiber <- restore(self).forkDaemon
-        a     <- restore(fiber.join).onInterrupt(fiber.interruptAs(id).forkDaemon)
+        a     <- fiber.join.interruptible.onInterrupt(fiber.interruptAs(id).forkDaemon)
```

Comment:
I don't think this is intuitive, this means that `x.disconnect` automatically implies `x.interruptible` for the current fiber - i.e. any critical section can be terminated if there's disconnect. I think the act of disconnecting the interruption of x should be separate from making sure this interruption can actually happen.

Reply:
I don't think that is right because we are still doing `restore(self)` on the underlying effect. Without this change I don't think `disconnect` has its intended effect of allowing you to continue without waiting for interruption to complete if called within an uninterruptible region.

Reply:
@adamgfraser I think it's logical that within an uninterruptible region `disconnect` becomes almost a no-op – because the outer effect can't receive interruptions, the underlying effect can't be interrupted in normal circumstances. _Forcing_ it to receive interruptions compromises the entire uninterruptible region of the outer fiber, I don't think that would make sense. And you can use `self.uninterruptible.disconnect.interruptible` to force the outer effect's `join` to be interruptible, while leaving the inner effect uninterruptible.

Note, _even_ with this change if the underlying effect is interrupted, but not the outer effect, the outer effect will still wait until the fiber stops completely before returning from `fiber.join` - observers aren't woken up until the fiber finishes completely.

So, to guarantee that interruption of the inner effect results in immediate resumption of the outer effect, we must use a Promise + onInterrupt in the inner effect – `fiber.join.interruptible` does not achieve this, AND it exposes the entire region to interruption which is avoidable with Promises.

Reply:
@neko-kai I added some tests to try to demonstrate what I think is the correct behavior.

I see the impact of `disconnect` in an uninterruptible region as quite significant, and essentially guaranteeing that interruption of a disconnected effect will return immediately regardless of whether the region is interruptible or uninterruptible.

This can be done in a way where we are only ensuring that the interruption can return immediately, the underlying effect itself is still guaranteed to execute if it is in an uninterruptible region.